gnomad_qc/v3/sample_qc/sample_qc.py

```diff
@@ -334,86 +556,369 @@ def apply_stratified_filters(filtering_qc_metrics: List[str]) -> hl.Table:
-        metric_threshold={'n_singleton_residual': (4.0, 8.0)}
+        metric_threshold={'n_singleton_residual': (math.inf, 8.0), 'r_het_hom_var_residual': (math.inf, 4.0)}
     )
 
     residuals_ht = residuals_ht.annotate(
-        **stratified_metrics_ht[sample_qc_ht.key]
+        **stratified_metrics_ht[residuals_ht.key]
     )
     residuals_ht = residuals_ht.annotate_globals(
-        **stratified_metrics_ht.index_globals()
+        **stratified_metrics_ht.index_globals(),
+        n_pcs=n_pcs,
     )
 
     return residuals_ht
 
 
-def generate_metadata() -> hl.Table:
-    meta_ht = project_meta.ht()
-    sample_qc_ht = get_sample_qc("bi_allelic").ht()
-    hard_filters_ht = hard_filtered_samples.ht()
-    regressed_metrics_ht = regressed_metrics.ht()
-    pop_ht = pop.ht()
-    release_related_samples_to_drop_ht = release_related_samples_to_drop.ht()
-    pca_related_samples_to_drop_ht = pca_related_samples_to_drop.ht()
-    sex_ht = sex.ht()
-    sex_ht = sex_ht.transmute(
+def get_relatedness_set_ht(relatedness_ht: hl.Table) -> hl.Table:
+    """
+    Parses relatedness Table to get every relationship (except UNRELATED) per sample.
+
+    Returns Table keyed by sample with all sample relationships in a set.
+
+    :param Table relatedness_ht: Table with inferred relationship information output by pc_relate.
+        Keyed by sample pair (i, j).
+    :return: Table keyed by sample (s) with all relationships annotated as a set.
+    :rtype: hl.Table
+    """
+    relatedness_ht = relatedness_ht.filter(relatedness_ht.relationship != UNRELATED)
+    relatedness_ht = relatedness_ht.select("relationship", s=relatedness_ht.i.s).union(
+        relatedness_ht.select("relationship", s=relatedness_ht.j.s)
+    )
+    relatedness_ht = relatedness_ht.group_by(relatedness_ht.s).aggregate(
+        relationships=hl.agg.collect_as_set(relatedness_ht.relationship)
+    )
+    return relatedness_ht
+
+
+def get_relationship_filter_expr(
+    hard_filtered_expr: hl.expr.BooleanExpression,
+    relationship: str,
+    relationship_set: hl.expr.SetExpression,
+) -> hl.expr.builders.CaseBuilder:
+    """
+    Returns case statement to populate relatedness filters in sample_filters struct
+
+    :param hl.expr.BooleanExpression hard_filtered_expr: Boolean for whether sample was hard filtered.
+    :param str relationship: Relationship to check for. One of DUPLICATE_OR_TWINS, PARENT_CHILD, or SIBLINGS.
+    :param hl.expr.SetExpression relationship_set: Set containing all possible relationship strings for sample.
+    :return: Case statement used to population sample_filters related filter field.
+    :rtype: hl.expr.builders.CaseBuilder
+    """
+    return (
+        hl.case()
+        .when(hard_filtered_expr, hl.null(hl.tbool))
+        .when(hl.is_defined(relationship_set), relationship_set.contains(relationship))
+        .default(False)
+    )
+
+
+def compare_row_counts(ht1: hl.Table, ht2: hl.Table) -> bool:
```

Comment:
Thoughts on eventually turning this into common code? It generic enough, although I don't know if we plan on comparing the sizes of two tables in the future (it might be useful as a sanity check).